Support both train / eval modes for ModuleInfo

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #78735

This PR enhances the tests in `test/test_modules.py` to be run across train / eval modes. It makes the following changes:
* Adds a required `training` arg to `ModuleInfo.module_inputs_func` - this allows those functions to generate sample module inputs based on the training mode.
* Updates the `@modules` decorator to additionally pass a `training` arg to test functions.
    * The `training` arg values to pass can be customized via `TrainEvalMode` e.g. `@modules(..., train_eval_mode=TrainEvalMode.train_only)`. Supported modes are `train_only`, `eval_only`, and `train_and_eval`.
    * Tests that use the decorator in `test/test_modules.py` have been updated. They now ingest the `training` arg, pass it to the `module_inputs_func`, and set the training mode for instantiated module instances with `m.train(training)`. There's almost certainly a way to do this with less boilerplate.
* Adds a new `train_and_eval_differ` option to `ModuleInfo` for indicating whether a specific module cares about the training mode.
    * If this is `False`, tests are generated just for a single mode. This avoids pointlessly running tests twice when the training mode setting doesn't matter.
    * Adds a new test in `test/test_modules.py` to verify that `train_and_eval_differ` is set correctly. It does this by instantiating each module, deleting the `training` attribute, and then running the forward pass. If the forward pass references `training`, it will throw an `AttributeError`, indicating that `train_and_eval_differ=True` should be set.

jbschlosser committed:
Support both train / eval modes for ModuleInfo

[ghstack-poisoned]

jbschlosser committed:
Update on "Support both train / eval modes for ModuleInfo"


This PR enhances the tests in `test/test_modules.py` to be run across train / eval modes. It allows sample inputs to be generated based on the mode under test (useful if e.g. if a certain sample should only be used during eval mode).

[ghstack-poisoned]

jbschlosser committed:
Update on "Support both train / eval modes for ModuleInfo"


This PR enhances the tests in `test/test_modules.py` to be run across train / eval modes. It makes the following changes:
* Adds a required `training` arg to `ModuleInfo.module_inputs_func` - this allows those functions to generate sample module inputs based on the training mode.
* Updates the `modules` decorator to additionally pass a `training` arg to test functions.
    * The `training` arg values to pass can be customized via `TrainEvalMode` e.g. `modules(..., train_eval_mode=TrainEvalMode.train_only)`. Supported modes are `train_only`, `eval_only`, and `train_and_eval`.
    * Tests that use the decorator in `test/test_modules.py` have been updated. They now ingest the `training` arg, pass it to the `module_inputs_func`, and set the training mode for instantiated module instances with `m.train(training)`. There's almost certainly a way to do this with less boilerplate.
* Adds a new `train_and_eval_differ` option to `ModuleInfo` for indicating whether a specific module cares about the training mode.
    * If this is `False`, tests are generated just for training by default. This avoids pointlessly running tests twice when the training mode setting doesn't matter.
    * Adds a new test in `test/test_modules.py` to verify that `train_and_eval_differ` is set correctly. It does this by instantiating each module, deleting the `training` attribute, and then running the forward pass. If the forward pass references `training`, it will throw an `AttributeError`, indicating that `train_and_eval_differ=True` should be set.

[ghstack-poisoned]

albanD reviewed (approved):
SGTM!
Just minor comments

jbschlosser commented on test/test_modules.py:
```diff
@@ -233,15 +241,15 @@ def test_pickle(self, device, dtype, module_info):
                     output_from_copy = m_copy(*args, **kwargs)
                     self.assertEqual(output, output_from_copy)
 
+    @skipMeta
```
Just reordered it slightly for consistency, before this was underneath `@modules`

albanD replied:
Ho ok

jbschlosser commented on test/test_modules.py:
```diff
@@ -640,6 +654,43 @@ def inner_check_out_mem_format(output):
+                                        "affect the forward pass. Consider setting train_and_eval_differ=True "
+                                        "for this ModuleInfo entry.")
+                    else:
+                        raise
```
That's my understanding, but I'll change it to `raise e` for clarity

jbschlosser commented on test/test_modules.py:
```diff
@@ -640,6 +654,43 @@ def inner_check_out_mem_format(output):
+            if module_input.forward_input is None:
+                continue
+
+            with freeze_rng_state():
```
Not strictly needed here; this is used in other tests for determinism. Removing

jbschlosser reviewed (commented):
@albanD Before I can merge this, it looks like I need a way to skip a test for eval mode only.

I have a working fix that can skip based on anything passed to the test - going to add a commit earlier in the PR stack that adds this

jbschlosser committed:
Update on "Support both train / eval modes for ModuleInfo"


This PR enhances the tests in `test/test_modules.py` to be run across train / eval modes. It makes the following changes:
* Adds a required `training` arg to `ModuleInfo.module_inputs_func` - this allows those functions to generate sample module inputs based on the training mode.
* Updates the `modules` decorator to additionally pass a `training` arg to test functions.
    * The `training` arg values to pass can be customized via `TrainEvalMode` e.g. `modules(..., train_eval_mode=TrainEvalMode.train_only)`. Supported modes are `train_only`, `eval_only`, and `train_and_eval`.
    * Tests that use the decorator in `test/test_modules.py` have been updated. They now ingest the `training` arg, pass it to the `module_inputs_func`, and set the training mode for instantiated module instances with `m.train(training)`. There's almost certainly a way to do this with less boilerplate.
* Adds a new `train_and_eval_differ` option to `ModuleInfo` for indicating whether a specific module cares about the training mode.
    * If this is `False`, tests are generated just for a single mode. This avoids pointlessly running tests twice when the training mode setting doesn't matter.
    * Adds a new test in `test/test_modules.py` to verify that `train_and_eval_differ` is set correctly. It does this by instantiating each module, deleting the `training` attribute, and then running the forward pass. If the forward pass references `training`, it will throw an `AttributeError`, indicating that `train_and_eval_differ=True` should be set.

[ghstack-poisoned]

malfet commented:
@pytorchbot revert -m "Broke eval tests on Win, 10.2 and ROCM, see https://hud.pytorch.org/pytorch/pytorch/commit/12658fcd5bdf4d2437754633b3fa39ab15d213b9 " -c missingsignal

malfet commented:
@pytorchbot revert -m "Broke eval tests on Win, 10.2 and ROCM, see https://hud.pytorch.org/pytorch/pytorch/commit/12658fcd5bdf4d2437754633b3fa39ab15d213b9 "  -c nosignal

jbschlosser committed:
Update on "Support both train / eval modes for ModuleInfo"


This PR enhances the tests in `test/test_modules.py` to be run across train / eval modes. It makes the following changes:
* Adds a required `training` arg to `ModuleInfo.module_inputs_func` - this allows those functions to generate sample module inputs based on the training mode.
* Updates the `modules` decorator to additionally pass a `training` arg to test functions.
    * The `training` arg values to pass can be customized via `TrainEvalMode` e.g. `modules(..., train_eval_mode=TrainEvalMode.train_only)`. Supported modes are `train_only`, `eval_only`, and `train_and_eval`.
    * Tests that use the decorator in `test/test_modules.py` have been updated. They now ingest the `training` arg, pass it to the `module_inputs_func`, and set the training mode for instantiated module instances with `m.train(training)`. There's almost certainly a way to do this with less boilerplate.
* Adds a new `train_and_eval_differ` option to `ModuleInfo` for indicating whether a specific module cares about the training mode.
    * If this is `False`, tests are generated just for a single mode. This avoids pointlessly running tests twice when the training mode setting doesn't matter.
    * Adds a new test in `test/test_modules.py` to verify that `train_and_eval_differ` is set correctly. It does this by instantiating each module, deleting the `training` attribute, and then running the forward pass. If the forward pass references `training`, it will throw an `AttributeError`, indicating that `train_and_eval_differ=True` should be set.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge -a
